deps/GraphBLAS/Source/GxB_Matrix_tuple_iter.c

```diff
@@ -13,31 +13,73 @@ static inline void _EmptyIterator
 	iter->nnz_idx = 0;
 }
 
+static inline void _MatrixTupleIter_init
+(
+	GxB_MatrixTupleIter *iter,      // iterator to init
+	GrB_Matrix A,                   // matrix to iterate over
+	int sparsity_type               // sparsity type of the matrix
+) {
+	GrB_Index nrows;
+	GrB_Matrix_nrows(&nrows, A) ;
+
+	GrB_Matrix_nvals(&iter->nvals, A) ;
+	iter->A = A ;
+	iter->nnz_idx = 0 ;
+	iter->row_idx = 0 ;
+	iter->nrows = nrows;
+	iter->p = A->p[0] ;
+	iter->sparsity_type = sparsity_type;
+	if(sparsity_type == GxB_HYPERSPARSE) {
+		iter->sparse_row_idx = 0;
+		iter->nrows_non_empty = A->nvec;
+	}
+}
+
 // Create a new iterator
 GrB_Info GxB_MatrixTupleIter_new
 (
 	GxB_MatrixTupleIter **iter,     // iterator to create
-	GrB_Matrix A                    // matrix to iterate over
+	GrB_Matrix A,                   // matrix to iterate over
+	int sparsity_type               // sparsity type of the matrix
 ) {
-	GB_WHERE(A, "GxB_MatrixTupleIter_new (A)") ;
+	GB_WHERE(A, "GxB_MatrixTupleIter_iterate_row (A)") ;
 	GB_RETURN_IF_NULL_OR_FAULTY(A) ;
 
 	// make sure matrix is not bitmap or full
-	GxB_set(A, GxB_SPARSITY_CONTROL, GxB_SPARSE) ;
-
-	GrB_Index nrows;
-	GrB_Matrix_nrows(&nrows, A) ;
+	GxB_set(A, GxB_SPARSITY_CONTROL, sparsity_type) ;
 
 	*iter = GB_MALLOC(1, GxB_MatrixTupleIter) ;
-	GrB_Matrix_nvals(&((*iter)->nvals), A) ;
-	(*iter)->A = A ;
-	(*iter)->nnz_idx = 0 ;
-	(*iter)->row_idx = 0 ;
-	(*iter)->nrows = nrows;
-	(*iter)->p = A->p[0] ;
+	_MatrixTupleIter_init(*iter, A, sparsity_type);
 	return (GrB_SUCCESS) ;
 }
 
+// finds the row index in Ah for HYPERSPARSE matrix,
+// returns true if found else false.
+static inline bool _find_row_index_in_Ah
+(
+	GxB_MatrixTupleIter *iter,
+	GrB_Index rowIdx,
+	GrB_Index *result
+) {
+	if(!result || !iter || !iter->nrows_non_empty) return false;
+	GrB_Index l = 0, h = iter->nrows_non_empty - 1, m;
+	// find the index using binary search
+	while(l <= h) {
+		m = (l + h)/2;
+		int64_t val = iter->A->h[m];
+		if(val == rowIdx) {
+			*result = m;
+			return true;
+		} else if(val < rowIdx) {
+			l = m + 1;
+		} else { // val > rowIdx
+			h = m - 1;
+		}
+	}
+
+	return false;
+}
+
```

Comment:
See if this function can be replaced by:
`GB_BINARY_SEARCH` from `deps/GraphBLAS/Source/GB_binary_search.h`